### PR TITLE
Added config parameter to WCA module, for allowing to export pre and …

### PR DESCRIPTION
**Pull request by massimo-ferraro**
…post contingencies and actions states during WCA processing
Fix in reading domains output

### CODE REVIEW

File: wca-integration/src/main/java/eu/itesla_project/wca/WCAConfig.java [hunk 1]
```diff
@@ -1,5 +1,6 @@
 /**
  * Copyright (c) 2016, All partners of the iTesla project (http://www.itesla-project.eu/consortium)
+ * Copyright (c) 2016, RTE (http://www.rte-france.com)
```
Comment (mathbagu) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAConfig.java]: 2017

File: wca-integration/src/main/java/eu/itesla_project/wca/WCAConfig.java [hunk 2]
```diff
@@ -25,20 +26,24 @@
...
         boolean debug = config.getBooleanProperty("debug", false);
-        return new WCAConfig(xpressHome, reducedVariableRatio, debug);
+        boolean exportStates = config.getBooleanProperty("exportStates", false);
+        return new WCAConfig(xpressHome, reducedVariableRatio, debug, exportStates);
     }
 
-    public WCAConfig(Path xpressHome, float reducedVariableRatio, boolean debug) {
+    public WCAConfig(Path xpressHome, float reducedVariableRatio, boolean debug, boolean exportStates) {
         Objects.requireNonNull(xpressHome);
         this.xpressHome = xpressHome;
```
Comment (mathbagu) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAConfig.java]: this.xpressHome = Objects.requireNonNull(xpressHome)

File: wca-integration/src/main/java/eu/itesla_project/wca/WCAImpl.java [hunk 3]
```diff
@@ -318,11 +318,15 @@ private static void copyRequired(Path workingDir) throws IOException {
 
                         // write base state
                         new AmplNetworkWriter(network, dataSource, 0, 0, false, mapper, CLUSTERS_AMPL_EXPORT_CONFIG).write();
+                        if (config.isExportStates())
+                            new WCAStateExporter(network, workingDir, 0, 0).export();
```
Comment (mathbagu) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAImpl.java]: Add braces to respect java coding guideline even if there is only one statement.

File: wca-integration/src/main/java/eu/itesla_project/wca/WCAImpl.java [hunk 4]
```diff
@@ -318,11 +318,15 @@ private static void copyRequired(Path workingDir) throws IOException {
...
 
                         // write post contingency state
                         network.getStateManager().setWorkingState(contingencyStateId);
                         AmplUtil.fillMapper(mapper, network); // because action can create a new bus
                         new AmplNetworkWriter(network, dataSource, contingencyNum, 0, true, mapper, CLUSTERS_AMPL_EXPORT_CONFIG).write();
+                        if (config.isExportStates())
+                            new WCAStateExporter(network, workingDir, contingencyNum, 0).export();
```
Comment (mathbagu) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAImpl.java]: Add braces to respect java coding guideline even if there is only one statement.

File: wca-integration/src/main/java/eu/itesla_project/wca/WCAImpl.java [hunk 5]
```diff
@@ -339,6 +343,8 @@ private static void copyRequired(Path workingDir) throws IOException {
                             network.getStateManager().setWorkingState(curativeStateId);
                             AmplUtil.fillMapper(mapper, network); // because action can create a new bus
                             new AmplNetworkWriter(network, dataSource, contingencyNum, curativeActionNum, true, mapper, CLUSTERS_AMPL_EXPORT_CONFIG).write();
+                            if (config.isExportStates())
+                                new WCAStateExporter(network, workingDir, contingencyNum, curativeActionNum).export();
```
Comment (mathbagu) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAImpl.java]: Add braces to respect java coding guideline even if there is only one statement.

File: wca-integration/src/main/java/eu/itesla_project/wca/WCAImpl.java [hunk 6]
```diff
@@ -402,6 +408,8 @@ public WCACluster after(Path workingDir, ExecutionReport report) throws IOExcept
 
                         // write base state
                         new AmplNetworkWriter(network, dataSource, 0, 0, false, mapper, DOMAINS_AMPL_EXPORT_CONFIG).write();
+                        if (config.isExportStates())
+                            new WCAStateExporter(network, workingDir, 0, 0).export();
```
Comment (mathbagu) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAImpl.java]: Add braces to respect java coding guideline even if there is only one statement.

File: wca-integration/src/main/java/eu/itesla_project/wca/WCAImpl.java [hunk 7]
```diff
@@ -415,6 +423,8 @@ public WCACluster after(Path workingDir, ExecutionReport report) throws IOExcept
                             network.getStateManager().setWorkingState(preventiveStateId);
                             AmplUtil.fillMapper(mapper, network); // because action can create a new bus
                             new AmplNetworkWriter(network, dataSource, 0, preventiveActionNum, true, mapper, DOMAINS_AMPL_EXPORT_CONFIG).write();
+                            if (config.isExportStates())
+                                new WCAStateExporter(network, workingDir, 0, preventiveActionNum).export();
```
Comment (mathbagu) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAImpl.java]: Add braces to respect java coding guideline even if there is only one statement.

File: wca-integration/src/main/java/eu/itesla_project/wca/WCAStateExporter.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    private final Network network;
+    private final Path folder;
+    private final int faultNum;
+    private final int actionNum;
+
+    public WCAStateExporter(Network network, Path folder, int faultNum, int actionNum) {
+        this.network = network;
+        this.folder = folder;
+        this.faultNum = faultNum;
+        this.actionNum = actionNum;
```
Comment (mathbagu) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAStateExporter.java]: Add assertion: network and folder should not be null.

File: wca-integration/src/main/java/eu/itesla_project/wca/WCAStateExporter.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+    public void export() {
+        Properties parameters = new Properties();
+        parameters.setProperty("iidm.export.xml.indent", "true");
+        parameters.setProperty("iidm.export.xml.with-branch-state-variables", "true");
+        DataSource dataSource = new GzFileDataSource(folder, network.getId() + "_" + faultNum + "_" + actionNum);
+        Exporters.export("XIIDM", network, parameters, dataSource);
+    }
+
+}
```
Comment (mathbagu) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAStateExporter.java]: This class could be replaced by an Interface with a static method "export"

File: wca-integration/src/main/java/eu/itesla_project/wca/WCAUtils.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import eu.itesla_project.iidm.network.Network;
+
+/**
+ *
+ * @author Massimo Ferraro <massimo.ferraro@techrain.it>
+ */
+public final class WCAUtils {
+
+    private WCAUtils() {
+    }
```
Comment (massimo-ferraro) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAUtils.java]: Would you like to have a new interface, with just this static method? But what would be the purpose of the interface (besides having the static method)? Who would/should implement it?
Or would you like to have this static method in an existing interface, WCA?

Comment (mathbagu) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAUtils.java]: I think that old utility classes should be converted to utility interfaces with static methods. In this specific case, there is no reason to create an instance (no internal data), so the exportState method is static but you have to declare this class as final and add a private constructor.
This pattern is equivalent to an interface with a single static method. This interface can be private package.

This is a suggestion: if you think your implementation is better, just keep it.

Comment (massimo-ferraro) [inline wca-integration/src/main/java/eu/itesla_project/wca/WCAUtils.java]: In this case, as there is no contract that needs to be defined through an interface,, even if they have an equivalent pattern, I think that a utility class is better suited than an interface

**Pull request merged**